src/Swift.Runtime/src/Swift/Data.cs

```diff
@@ -1,10 +1,119 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using Swift;
+using Swift.Runtime;
+
+
 /// <summary>
-/// Represents Foundation.Data type.
+/// Represents Swift Foundation.DataProtocol in C#.
+/// </summary>
+public interface ISwiftDataProtocol { }
+
+/// <summary>
+/// Represents Swift Foundation.ContiguousBytes in C#.
 /// </summary>
+public interface ISwiftContiguousBytes { }
 
-// TODO: https://github.com/dotnet/runtimelab/issues/2992
-public struct Data
+/// <summary>
+/// Represents Foundation.Data type.
+/// </summary>
+public struct Data : ISwiftObject
 {
     private long _flags;
     private IntPtr _object;
+
+    private static nuint _payloadSize = SwiftObjectHelper<Data>.GetTypeMetadata().Size;
+
+    private static Dictionary<Type, string> _protocolConformanceSymbols;
+
+    static Data()
+    {
+        _protocolConformanceSymbols = new Dictionary<Type, string> {
+            { typeof(ISwiftDataProtocol), "$s10Foundation4DataVAA0B8ProtocolAAMc" },
+            { typeof(ISwiftContiguousBytes), "$s10Foundation4DataVAA15ContiguousBytesAAMc" },
+        };
```

Comment:
Side note (definitely not this PR): This induces a cost to just touching the Data type - it's basically a startup cost. But there's no good reason for it - this is data which is hardcoded on the type. What if we encoded this in "code". Basically instead generating code which fills dictionary - generate a switch statement. For smallish number of cases the perf is probably better for switch anyway and there's 0 startup cost (with AOT it's true 0, with JIT it's as lazy as it can be).

Reply:
Good point, thanks. Added to https://github.com/dotnet/runtimelab/issues/2990.